src/transport/tests/TestPeerMessageCounter.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+void GroupBackTrackTest(nlTestSuite * inSuite, void * inContext)
+{
+    for (auto n : counterValuesArray)
+    {
+        chip::Transport::PeerMessageCounter counter;
+        NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n, true) == CHIP_NO_ERROR);
+
+        counter.CommitWithRollOver(n);
+        // 1.   Some set of values N - k come in, for 0 < k < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW.
+        //      All of those should be considered valid and committed.
+        for (uint32_t k = 1; k * k < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE; k++)
+        {
+            NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n - (k * k), true) == CHIP_NO_ERROR);
+            counter.CommitWithRollOver(n - (k * k));
+        }
+        // 2. Counter value N + 3 comes in
+        NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n + 3, true) == CHIP_NO_ERROR);
+        counter.CommitWithRollOver(n + 3);
+
+        // 3. The same set of values N - k come in as in step (1) and are all considered duplicates/out of window.
+        for (uint32_t k = 1; k * k < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE; k++)
+        {
+            NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n - (k * k), true) != CHIP_NO_ERROR);
+        }
+
+        // 4. The values that were not in the set in step (a) (but are at least N + 3 - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW)
+        //    come in, and all are treated as allowed.
+        for (uint32_t k = n + 3 - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE; k != n + 3; ++k)
+        {
+            if (n - k == 0 || n - k == 1 || n - k == 4 || n - k == 9 || n - k == 16 || n - k == 25)
+            {
+                continue;
+            }
+            NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(k, true) == CHIP_NO_ERROR);
+            counter.CommitWithRollOver(k);
+        }
+    }
+}
+
+void GroupBigLeapTest(nlTestSuite * inSuite, void * inContext)
+{
+    for (auto n : counterValuesArray)
+    {
+        for (uint32_t k = (static_cast<uint32_t>(1 << 31) - 5); k <= (static_cast<uint32_t>(1 << 31) - 1); k++)
+        {
+            chip::Transport::PeerMessageCounter counter;
+            NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n, true) == CHIP_NO_ERROR);
+
+            counter.CommitWithRollOver(n);
+
+            // 1. A counter value of N + k comes in, we detect it as valid and commit it.
+            NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n + k, true) == CHIP_NO_ERROR);
+            counter.CommitWithRollOver(n + k);
+
+            // 2. A counter value of N comes in, we detect it as duplicate.
+            NL_TEST_ASSERT(inSuite, counter.VerifyOrTrustFirst(n, true) != CHIP_NO_ERROR);
+
+            // 3. A counter value between N - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW and N + k - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW
+            //    (but not including N + k - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW) comes in, we treat it as duplicate.
+
+            // Only test some values to save processing time
```

Comment:
Ah, good catch. This part of my suggested test just makes no sense (and in particular `N - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW` should be a valid value here, as your code below says).

I like your approach here.

Reply:
Thanks